envision/server.py

```diff
@@ -54,30 +64,214 @@ def options(self):
         self.finish()
 
 
+class Frame:
+    def __init__(self, data: str, timestamp: int, next_=None):
+        # Time since epoch in seconds
+        self._timestamp = timestamp
+        self._data = data
+        self._size = sys.getsizeof(data)
+        self.next_ = next_
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, data: str):
+        self._data = data
+        self._size = sys.getsizeof(data)
+
+    @property
+    def size(self):
+        return self._size
+
+
+class Frames:
+    def __init__(self, max_capacity_mb=500):
+        self._origin_timestamp = None
+        self._max_capacity = max_capacity_mb
+
+        # XXX: Index of timestamp in `self._timestamps` matches the index of the
+        # frame with the same timestamp in `self._frames`.
+        self._timestamps = []
+        self._frames = []
+
+    @property
+    def start_frame(self):
+        if len(self._frames) == 0:
+            return None
+
+        return self._frames[0]
+
+    @property
+    def start_time(self):
+        return self._origin_timestamp
+
+    @property
+    def elapsed_time(self):
+        return self._frames[-1].timestamp - self._origin_timestamp
+
+    def append(self, frame: Frame):
+        if len(self._frames) == 0:
+            self._origin_timestamp = frame.timestamp
+
+        self._enforce_max_capacity()
+
+        if len(self._frames) >= 1:
+            self._frames[-1].next_ = frame
+
+        self._frames.append(frame)
+        self._timestamps.append(frame.timestamp)
+
+    def __call__(self, timestamp):
+        """Finds the nearest frame according to the given timestamp."""
+        frame_idx = bisect.bisect_left(self._timestamps, timestamp)
+        if frame_idx >= len(self._frames):
+            frame_idx = -1
+
+        return self._frames[frame_idx]
+
+    def _enforce_max_capacity(self):
```

Comment:
Should be cautious and conservative about memory management. The max capacity is applied to every simulation, and running a loop scenario could generate 500M data in less than 10 mins.